pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/periodictask/ControllerPeriodicTask.java

```diff
@@ -55,14 +57,25 @@ public ControllerPeriodicTask(String taskName, long runFrequencyInSeconds, long
   }
 
   @Override
-  protected final void runTask() {
+  protected final void runTask(Properties periodicTaskProperties) {
     _controllerMetrics.addMeteredTableValue(_taskName, ControllerMeter.CONTROLLER_PERIODIC_TASK_RUN, 1L);
     try {
+      // Check if we have a specific table against which this task needs to be run.
+      String propTableNameWithType = (String) periodicTaskProperties.get(PeriodicTask.PROPERTY_KEY_TABLE_NAME);
+
       // Process the tables that are managed by this controller
       List<String> tablesToProcess = new ArrayList<>();
-      for (String tableNameWithType : _pinotHelixResourceManager.getAllTables()) {
-        if (_leadControllerManager.isLeaderForTable(tableNameWithType)) {
-          tablesToProcess.add(tableNameWithType);
+      if (propTableNameWithType == null){
+        // Table name is not available, so task should run on all tables for which this controller is the lead.
+        for (String tableNameWithType : _pinotHelixResourceManager.getAllTables()) {
+          if (_leadControllerManager.isLeaderForTable(tableNameWithType)) {
+            tablesToProcess.add(tableNameWithType);
+          }
+        }
+      } else {
+        // Table name is available, so task should run only on the specified table.
+        if (_leadControllerManager.isLeaderForTable(propTableNameWithType)) {
+          tablesToProcess.add(propTableNameWithType);
         }
       }
       processTables(tablesToProcess);
```

Comment:
Only call `processTables` when `tablesToProcess` is not empty to avoid the overhead

Reply:
Fixed.